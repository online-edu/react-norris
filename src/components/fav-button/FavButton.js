import React, { Component } from 'react';
import PropTypes from 'prop-types';

/**
 * FavButton component.
 */
class FavButton extends Component {
  /**
   * FavButton props types.
   */
  static get propTypes() {
    return {
      /** Get fav ot not fav flag */
      favorite: PropTypes.bool.isRequired,
      /** Gets called when the fav button is clicked. */
      onClick: PropTypes.func.isRequired,
    };
  }

  /**
   * Intilalizes the state and binds all methods.
   *
   * @param {Object} props - access 'props'
   */
  constructor(props) {
    super(props);
    const { favorite } = this.props;
    this.toggle = favorite;
    this.state = {
      icon: favorite ? 'favorite' : 'favorite_border',
    };
    this.onToggle = this.onToggle.bind(this);
  }

  /**
   * Toggle icon for button and notify parent.
   */
  onToggle() {
    this.setState(
      {
        icon: this.toggle ? 'favorite_border' : 'favorite',
      },
      () => {
        this.toggle = !this.toggle;
        const { onClick } = this.props;
        onClick(this.toggle);
      },
    );
  }

  /**
   * Render method for component
   */
  render() {
    const { icon } = this.state;
    return (
      <button type="button" className="btn btn-link" onClick={this.onToggle}>
        <i className="material-icons pt-2">{icon}</i>
      </button>
    );
  }
}

export default FavButton;
