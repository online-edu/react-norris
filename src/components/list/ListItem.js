import React from 'react';
import PropTypes from 'prop-types';
import FavButton from '../fav-button';

/**
 * List item component.
 */
const ListItem = ({ favorite, item, onFavoriteClick }) => (
  <li className="list-group-item d-flex justify-content-between align-items-center">
    {item.joke}
    <FavButton
      favorite={favorite}
      onClick={toggle => onFavoriteClick(toggle)}
    />
  </li>
);
/**
 * List item props types.
 */
ListItem.propTypes = {
  /** Toggle favorite */
  favorite: PropTypes.bool.isRequired,
  /** Gets called when the fav button is clicked. */
  onFavoriteClick: PropTypes.func.isRequired,
  /** Movie details */
  item: PropTypes.shape({ joke: PropTypes.string.isRequired }).isRequired,
};

export default ListItem;
