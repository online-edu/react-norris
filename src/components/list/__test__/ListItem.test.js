import React from 'react';
import { shallow } from 'enzyme';
import ListItem from '../ListItem';
import FavButton from '../../fav-button';

describe('<ListItem />', () => {
  const onFavClick = jest.fn();
  const joke = {
    id: 480,
    joke: 'The class object inherits from Chuck Norris',
    categories: ['nerdy'],
  };

  const wrapper = (
    <ListItem item={joke} favorite onFavoriteClick={onFavClick} />
  );

  it('renders correctly', () => {
    const component = shallow(wrapper);
    expect(component).toMatchSnapshot();
  });

  it('contains class name "list-group-item" ', () => {
    const component = shallow(wrapper);
    expect(component.find('.list-group-item')).toHaveLength(1);
  });

  it('contains a FavButton ', () => {
    const component = shallow(wrapper);
    expect(component.find(FavButton)).toHaveLength(1);
  });

  it('simulates click events', () => {
    const component = shallow(wrapper);
    component.find(FavButton).simulate('click');
    expect(onFavClick).toHaveBeenCalledTimes(1);
  });
});
