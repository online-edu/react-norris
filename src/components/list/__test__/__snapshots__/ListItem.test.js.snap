// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ListItem /> renders correctly 1`] = `ShallowWrapper {}`;
